// n8n-workflow/dist/cjs/index.js
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./logger-proxy", "./node-helpers", "./observable-object", "./telemetry-helpers", "./errors", "./constants", "./common", "./cron", "./data-store.types", "./deferred-promise", "./global-state", "./interfaces", "./message-event-bus", "./execution-status", "./expression", "./expressions/expression-helpers", "./from-ai-parse-utils", "./node-helpers", "./tool-helpers", "./node-reference-parser-utils", "./metadata-utils", "./workflow", "./workflow-data-proxy", "./workflow-data-proxy-env-provider", "./versioned-node-type", "./type-validation", "./result", "./utils", "./type-guards", "./graph/graph-utils", "./extensions", "./extensions/expression-parser", "./native-methods", "./node-parameters/filter-parameter", "./node-parameters/parameter-type-validation", "./node-parameters/path-utils", "./evaluation-helpers"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.NativeMethods = exports.ExpressionParser = exports.ExpressionExtensions = exports.buildAdjacencyList = exports.parseExtractableSubgraphSelection = exports.isFilterValue = exports.isResourceLocatorValue = exports.isResourceMapperValue = exports.isINodePropertyOptionsList = exports.isINodePropertyCollectionList = exports.isINodePropertiesList = exports.isINodePropertyCollection = exports.isINodePropertyOptions = exports.isINodeProperties = exports.isDomainAllowed = exports.setSafeObjectProperty = exports.isSafeObjectProperty = exports.randomString = exports.randomInt = exports.updateDisplayOptions = exports.removeCircularRefs = exports.assert = exports.fileTypeFromMimeType = exports.sleepWithAbort = exports.sleep = exports.replaceCircularReferences = exports.jsonStringify = exports.base64DecodeUTF8 = exports.jsonParse = exports.deepCopy = exports.isObjectEmpty = exports.TelemetryHelpers = exports.ObservableObject = exports.NodeHelpers = exports.LoggerProxy = void 0;
    const LoggerProxy = __importStar(require("./logger-proxy"));
    exports.LoggerProxy = LoggerProxy;
    const NodeHelpers = __importStar(require("./node-helpers"));
    exports.NodeHelpers = NodeHelpers;
    const ObservableObject = __importStar(require("./observable-object"));
    exports.ObservableObject = ObservableObject;
    const TelemetryHelpers = __importStar(require("./telemetry-helpers"));
    exports.TelemetryHelpers = TelemetryHelpers;
    __exportStar(require("./errors"), exports);
    __exportStar(require("./constants"), exports);
    __exportStar(require("./common"), exports);
    __exportStar(require("./cron"), exports);
    __exportStar(require("./data-store.types"), exports);
    __exportStar(require("./deferred-promise"), exports);
    __exportStar(require("./global-state"), exports);
    __exportStar(require("./interfaces"), exports);
    __exportStar(require("./message-event-bus"), exports);
    __exportStar(require("./execution-status"), exports);
    __exportStar(require("./expression"), exports);
    __exportStar(require("./expressions/expression-helpers"), exports);
    __exportStar(require("./from-ai-parse-utils"), exports);
    __exportStar(require("./node-helpers"), exports);
    __exportStar(require("./tool-helpers"), exports);
    __exportStar(require("./node-reference-parser-utils"), exports);
    __exportStar(require("./metadata-utils"), exports);
    __exportStar(require("./workflow"), exports);
    __exportStar(require("./workflow-data-proxy"), exports);
    __exportStar(require("./workflow-data-proxy-env-provider"), exports);
    __exportStar(require("./versioned-node-type"), exports);
    __exportStar(require("./type-validation"), exports);
    __exportStar(require("./result"), exports);
    var utils_1 = require("./utils");
    Object.defineProperty(exports, "isObjectEmpty", { enumerable: true, get: function () { return utils_1.isObjectEmpty; } });
    Object.defineProperty(exports, "deepCopy", { enumerable: true, get: function () { return utils_1.deepCopy; } });
    Object.defineProperty(exports, "jsonParse", { enumerable: true, get: function () { return utils_1.jsonParse; } });
    Object.defineProperty(exports, "base64DecodeUTF8", { enumerable: true, get: function () { return utils_1.base64DecodeUTF8; } });
    Object.defineProperty(exports, "jsonStringify", { enumerable: true, get: function () { return utils_1.jsonStringify; } });
    Object.defineProperty(exports, "replaceCircularReferences", { enumerable: true, get: function () { return utils_1.replaceCircularReferences; } });
    Object.defineProperty(exports, "sleep", { enumerable: true, get: function () { return utils_1.sleep; } });
    Object.defineProperty(exports, "sleepWithAbort", { enumerable: true, get: function () { return utils_1.sleepWithAbort; } });
    Object.defineProperty(exports, "fileTypeFromMimeType", { enumerable: true, get: function () { return utils_1.fileTypeFromMimeType; } });
    Object.defineProperty(exports, "assert", { enumerable: true, get: function () { return utils_1.assert; } });
    Object.defineProperty(exports, "removeCircularRefs", { enumerable: true, get: function () { return utils_1.removeCircularRefs; } });
    Object.defineProperty(exports, "updateDisplayOptions", { enumerable: true, get: function () { return utils_1.updateDisplayOptions; } });
    Object.defineProperty(exports, "randomInt", { enumerable: true, get: function () { return utils_1.randomInt; } });
    Object.defineProperty(exports, "randomString", { enumerable: true, get: function () { return utils_1.randomString; } });
    Object.defineProperty(exports, "isSafeObjectProperty", { enumerable: true, get: function () { return utils_1.isSafeObjectProperty; } });
    Object.defineProperty(exports, "setSafeObjectProperty", { enumerable: true, get: function () { return utils_1.setSafeObjectProperty; } });
    Object.defineProperty(exports, "isDomainAllowed", { enumerable: true, get: function () { return utils_1.isDomainAllowed; } });
    var type_guards_1 = require("./type-guards");
    Object.defineProperty(exports, "isINodeProperties", { enumerable: true, get: function () { return type_guards_1.isINodeProperties; } });
    Object.defineProperty(exports, "isINodePropertyOptions", { enumerable: true, get: function () { return type_guards_1.isINodePropertyOptions; } });
    Object.defineProperty(exports, "isINodePropertyCollection", { enumerable: true, get: function () { return type_guards_1.isINodePropertyCollection; } });
    Object.defineProperty(exports, "isINodePropertiesList", { enumerable: true, get: function () { return type_guards_1.isINodePropertiesList; } });
    Object.defineProperty(exports, "isINodePropertyCollectionList", { enumerable: true, get: function () { return type_guards_1.isINodePropertyCollectionList; } });
    Object.defineProperty(exports, "isINodePropertyOptionsList", { enumerable: true, get: function () { return type_guards_1.isINodePropertyOptionsList; } });
    Object.defineProperty(exports, "isResourceMapperValue", { enumerable: true, get: function () { return type_guards_1.isResourceMapperValue; } });
    Object.defineProperty(exports, "isResourceLocatorValue", { enumerable: true, get: function () { return type_guards_1.isResourceLocatorValue; } });
    Object.defineProperty(exports, "isFilterValue", { enumerable: true, get: function () { return type_guards_1.isFilterValue; } });
    var graph_utils_1 = require("./graph/graph-utils");
    Object.defineProperty(exports, "parseExtractableSubgraphSelection", { enumerable: true, get: function () { return graph_utils_1.parseExtractableSubgraphSelection; } });
    Object.defineProperty(exports, "buildAdjacencyList", { enumerable: true, get: function () { return graph_utils_1.buildAdjacencyList; } });
    var extensions_1 = require("./extensions");
    Object.defineProperty(exports, "ExpressionExtensions", { enumerable: true, get: function () { return extensions_1.ExpressionExtensions; } });
    exports.ExpressionParser = __importStar(require("./extensions/expression-parser"));
    var native_methods_1 = require("./native-methods");
    Object.defineProperty(exports, "NativeMethods", { enumerable: true, get: function () { return native_methods_1.NativeMethods; } });
    __exportStar(require("./node-parameters/filter-parameter"), exports);
    __exportStar(require("./node-parameters/parameter-type-validation"), exports);
    __exportStar(require("./node-parameters/path-utils"), exports);
    __exportStar(require("./evaluation-helpers"), exports);
});
